source/files/dico/dictionnary.c

```diff
@@ -1,124 +1,120 @@
-    while(*scan != NULL)
-    {
-        if(b != NULL && b->data < (*scan)->data)
-        {
-            Dico* temp = b;
-            b = b->left;
-            temp->left = *scan;
-            *scan = temp;
-        }
-        scan = &(*scan)->left;
+int depth(Dico* tree){
+    if(tree == NULL){
+        return 0;//Attention définition. -1 ou 0
     }
-    if(b != NULL)
-    {
-        *scan = b;
+    else{
+        int depth_left = depth(tree->left);
+        int depth_right = depth(tree->right);
+        if(depth_left > depth_right){
+            return 1 + depth_left;
+        }
+        else{
+            return 1 + depth_right;
+        }
     }
-    return a;
 }
 
-Dico* all_left_dico(Tree* huff, char* buffer, int lsize, int psize)
-{
-    if(huff == NULL){return NULL;}
-    if(huff->data != 0)
-    {
-        Dico* ret = (Dico*)malloc(sizeof(Dico));
-        ret->key = huff->data;
-        ret->value = copy(buffer, lsize);
-        ret->left = NULL;
-        ret->right = NULL;
-
-        return ret;
+int bf(Dico* tree){
+    if(tree == NULL) {
+        return 0;
     }
-
-    int fr = 0;
-    if(lsize + 1 > psize)
-    {
-        psize += 100;
-        buffer = (char*)malloc(psize*sizeof(char));
-        fr = 1;
+    else{
+        return depth(tree->right) - depth(tree->left);
     }
-
-    buffer[lsize] = '0';
-    Dico* zero = all_left_dico(huff->left, buffer, lsize+1, psize);
-    buffer[lsize] = '1';
-    Dico* one = all_left_dico(huff->right, buffer, lsize+1, psize);
-
-    if(fr){free(buffer);}
-
-    return merge(zero, one);
 }
 
-int lenght(Dico* d)
-{
-    if(d != NULL)
-    {
-        return 1 + lenght(d->left);
+void right_rotation(Dico** tree){
+    if (*tree != NULL){
+        Dico* temp = (*tree)->left;
+        (*tree)->left = temp->right;
+        temp->right = *tree;
+        *tree = temp;
     }
-    return 0;
 }
-Dico* mid_of(Dico* d)
-{
-    int len = lenght(d);
 
-    Dico* ret = d;
-    for(int i = 0; i<len/2 - 1; i++)
-    {
-        ret = ret->left;
+void left_rotation(Dico** tree){
+    if (*tree != NULL){
+        Dico* temp = (*tree)->right;
+        (*tree)->right = temp->left;
+        temp->left = *tree;
+        *tree = temp;
     }
-    Dico* buffer = ret;
-    ret = ret->left;
-    buffer->left = NULL;
+}
 
-    return ret;
+void balance(Dico** tree){
+    if (*tree != NULL){
+        balance(&((*tree)->left));// Postfix
+        balance(&((*tree)->right));
+
+        int balance_factor = bf(*tree);
+        if (balance_factor <= -2){// Cas Gauche - ??
+            if(bf((*tree)->left) > 0){// Gauche - Droite
+                left_rotation(&((*tree)->left));
+            }
+            right_rotation(tree);// Gauche - Gauche
+        }
+        else if (balance_factor >= 2){// Cas Droite - ??
+            if(bf((*tree)->right) < 0){// Droite - Gauche
+                right_rotation(&((*tree)->right));
+            }
+            left_rotation(tree);// Droite - Droite
+        }
+    }
```

Comment:
I understand your copy-pasting, at least if you don't want to translate delete the comments